uni-snd/src/native_snd.rs

```diff
@@ -0,0 +1,111 @@
+use std;
+use std::thread;
+use std::sync::mpsc::{channel, Sender};
+
+use cpal;
+use uni_app::App;
+use super::SoundGenerator;
+
+pub struct SoundDriver<T: Send + 'static> {
+    event_loop: Option<cpal::EventLoop>,
+    format: Option<cpal::Format>,
+    stream_id: Option<cpal::StreamId>,
+    tx: Option<Sender<T>>,
+    generator: Option<Box<SoundGenerator<T>>>,
+}
+
+impl<T: Send + 'static> SoundDriver<T> {
+    pub fn new(generator: Box<SoundGenerator<T>>) -> Self {
+        let device = cpal::default_output_device();
+        let mut event_loop = None;
+        let mut format = None;
+        let mut stream_id = None;
+        if let Some(ref dev) = device {
+            match dev.default_output_format() {
+                Ok(fmt) => {
+                    let evt = cpal::EventLoop::new();
+                    match evt.build_output_stream(dev, &fmt) {
+                        Ok(str) => stream_id = Some(str),
+                        Err(e) => {
+                            App::print(format!("error : could not build output stream : {}\n", e))
+                        }
+                    }
+                    App::print(format!(
+                        "sound device : {} format {:?}\n",
+                        dev.name(),
+                        fmt.clone()
+                    ));
+                    format = Some(fmt);
+                    event_loop = Some(evt);
+                }
+                Err(e) => App::print(format!(
+                    "error : could not get default output format : {:?}\n",
+                    e
+                )),
+            }
+        } else {
+            App::print("warning : no sound device detected\n");
```

Comment:
What really happened with no sound device detected ? Will it still work silently ? If not, should we use `panic` here or return `Result<Self>` ?

Reply:
I would rather have it working silently with proper warning/errors in the console. Most games can still be played without sound. The idea right now is that the upper layer doesn't have to care if there is a sound device or not. All the functions always work but simply do nothing when no sound device is detected

Reply:
Okay, i understand, that's fine for me.